File: proposed-projects/round-trip-experiment/reverse/20260909T132655Z_2a16d72b/20260910T075017Z_244b957c/ported_implementation/src/grammar-graph/graph.ts
import { InputParseError } from '../utils/errors/input-parse-error.js';
import { isPointInRange } from '../utils/is-point-in-range.js';
import { Colorize, colorize, noColor } from './colorize.js';
import { GenericSet } from './generic-set.js';
import { getInputAsCodePoints } from './get-input-as-code-points.js';
import { getSerializedRuleKey } from './get-serialized-rule-key.js';
import { GraphNode, GraphNodeMeta } from './graph-node.js';
import { GraphPointer } from './graph-pointer.js';
import {
  isRange,
  isRuleChar,
  isRuleCharExcluded,
  isRuleEnd,
  isRuleRef,
} from './type-guards.js';
import type { RuleRef } from './rule-ref.js';
import type { Range, Rule, ValidInput } from './types.js';

type RootNode = Map<number, GraphNode>;
type Pointers = GenericSet<GraphPointer, string>;

const makePointers = (): Pointers =>
  new GenericSet<GraphPointer, string>((p) => p.id);

export class Graph {
  grammar: string;
  roots = new Map<number, RootNode>();
  previousCodePoints: number[] = [];
  private _rootNode: RootNode;

  constructor(
    grammar: string,
    stackedRules: (Rule | RuleRef)[][][],
    rootId: number
  ) {
    this.grammar = grammar;

    const ruleRefs: RuleRef[] = [];
    const uniqueRules = new GenericSet<Rule | RuleRef, string>(
      getSerializedRuleKey
    );
    for (let stackId = 0; stackId < stackedRules.length; stackId++) {
      const stack = stackedRules[stackId] ?? [];
      const nodes: RootNode = new Map();
      for (let pathId = 0; pathId < stack.length; pathId++) {
        const path = stack[pathId];
        let node: GraphNode | undefined = undefined;
        for (let stepId = path.length - 1; stepId >= 0; stepId--) {
          const nextNode = node;
          const rule = stack[pathId][stepId];
          uniqueRules.add(rule);
          if (isRuleRef(rule)) {
            ruleRefs.push(rule);
          }
          // rules coming in may be identical but have different references.
          // here, we ensure we always use the same reference for an identical
          // rule. this makes future comparisons easier.
          const uniqueRule = uniqueRules.get(rule);
          if (uniqueRule === undefined) {
            throw new Error('Could not get unique rule');
          }
          node = new GraphNode(
            uniqueRule,
            meta(stackId, pathId, stepId),
            nextNode
          );
        }
        if (node === undefined) {
          throw new Error('Could not get node');
        }
        nodes.set(pathId, node);
      }

      this.roots.set(stackId, nodes);
    }

    const rootNode = this.roots.get(rootId);
    if (!rootNode || rootNode.size === 0) {
      throw new Error(`Root node not found for value: ${rootId}`);
    }
    this._rootNode = rootNode;

    for (const ruleRef of ruleRefs) {
      const referencedNodes: GraphNode[] = [];
      for (const node of this.getRootNode(ruleRef.value).values()) {
        if (!referencedNodes.includes(node)) {
          referencedNodes.push(node);
        }
      }
      ruleRef.nodes = referencedNodes;
    }
  }

  private getRootNode(value: number): RootNode {
    const rootNode = this.roots.get(value);
    if (!rootNode || rootNode.size === 0) {
      throw new Error(`Root node not found for value: ${value}`);
    }
    return rootNode;
  }

  private getInitialPointers(): Pointers {
    const pointers = makePointers();

    const rootNode = this._rootNode;
    if (!rootNode) {
      throw new Error('Root node is not defined');
    }

    for (const [node, parent] of this.fetchNodesForRootNode(rootNode)) {
      const pointer = new GraphPointer(node, parent);
      for (const resolvedPointer of this.resolvePointer(pointer)) {
        pointers.add(resolvedPointer);
      }
    }
    return pointers;
  }

  private static setValid(pointers: GraphPointer[], valid: boolean): void {
    for (const pointer of pointers) {
      pointer.valid = valid;
    }
  }

  private parse(currentPointers: Pointers, codePoint: number): Pointers {
    for (const [rule, graphPointers] of Graph.iterateOverPointers(
      currentPointers
    )) {
      if (isRuleChar(rule)) {
        let valid = false;
        for (const possibleCodePoint of rule.value) {
          if (valid) {
            break;
          }
          if (isRange(possibleCodePoint)) {
            valid = isPointInRange(codePoint, possibleCodePoint as Range);
          } else {
            valid = codePoint === possibleCodePoint;
          }
        }
        Graph.setValid(graphPointers, valid);
      } else if (isRuleCharExcluded(rule)) {
        let valid = true;
        for (const possibleCodePoint of rule.value) {
          if (!valid) {
            break;
          }
          if (isRange(possibleCodePoint)) {
            valid = !isPointInRange(codePoint, possibleCodePoint as Range);
          } else {
            valid = codePoint !== possibleCodePoint;
          }
        }
        Graph.setValid(graphPointers, valid);
      } else if (!isRuleEnd(rule)) {
        throw new Error(`Unsupported rule: ${rule}`);
      }
    }

    // a pointer's id is the sum of its node's id and its parent's id chain.
    // if two pointers share the same id, it means they point to the same node and
    // have identical parent chains. for the purposes of walking the graph, we only
    // need to keep one of them.
    const nextPointers = makePointers();
    for (const currentPointer of currentPointers) {
      for (const unresolvedNextPointer of currentPointer.fetchNext()) {
        for (const resolvedNextPointer of this.resolvePointer(
          unresolvedNextPointer
        )) {
          nextPointers.add(resolvedNextPointer);
        }
      }
    }
    return nextPointers;
  }

  private *resolvePointer(
    unresolvedPointer: GraphPointer
  ): Generator<GraphPointer> {
    for (const resolvedPointer of unresolvedPointer.resolve()) {
      if (isRuleRef(resolvedPointer.node.rule)) {
        throw new Error(
          'Encountered a reference rule when building pointers to the graph'
        );
      }
      if (isRuleEnd(resolvedPointer.node.rule) && resolvedPointer.parent) {
        throw new Error(
          'Encountered an ending rule with a parent when building pointers to the graph'
        );
      }
      yield resolvedPointer;
    }
  }

  add(src: ValidInput, _pointers?: Pointers): Pointers {
    let pointers = _pointers !== undefined ? _pointers : this.getInitialPointers();
    const codePoints = getInputAsCodePoints(src);
    for (
      let codePointPos = 0;
      codePointPos < codePoints.length;
      codePointPos++
    ) {
      pointers = this.parse(pointers, codePoints[codePointPos]);
      if (pointers.size === 0) {
        throw new InputParseError(codePoints, codePointPos, [
          ...this.previousCodePoints,
        ]);
      }
    }
    for (const codePoint of codePoints) {
      this.previousCodePoints.push(codePoint);
    }
    return pointers;
  }

  /**
   * Yield either the node, or if a reference rule, the referenced node.
   *
   * This is distinct from the logic in GraphPointer because that needs a rule ref
   * with already defined nodes; this function is used to _set_ those nodes.
   */
  private *fetchNodesForRootNode(
    rootNodes: RootNode,
    parent?: GraphPointer
  ): Generator<[GraphNode, GraphPointer | undefined]> {
    for (const node of rootNodes.values()) {
      if (isRuleRef(node.rule)) {
        yield* this.fetchNodesForRootNode(
          this.getRootNode(node.rule.value),
          new GraphPointer(node, parent)
        );
      } else {
        yield [node, parent];
      }
    }
  }

  print(pointers?: Pointers, colors = false): string {
    const col: Colorize = colors ? colorize : noColor;
    const graphView: string[] = [];
    for (const nodes of this.roots.values()) {
      for (const node of nodes.values()) {
        graphView.push(node.print(pointers, true, col));
      }
    }
    return `\n${graphView.join('\n')}`;
  }

  toString(): string {
    return this.print(undefined, true);
  }

  private static *iterateOverPointers(
    pointers: Pointers
  ): Generator<[Rule, GraphPointer[]]> {
    // Rules are deduplicated by reference when the graph is built, so grouping by
    // object identity keeps distinct-but-equal rules apart the way a Map does.
    const seenRules = new Map<Rule, GraphPointer[]>();
    for (const pointer of pointers) {
      const rule = pointer.rule;
      if (isRuleRef(rule)) {
        throw new Error(
          'Encountered a reference rule in the graph, this should not happen'
        );
      }
      const entry = seenRules.get(rule as Rule);
      if (entry === undefined) {
        seenRules.set(rule as Rule, [pointer]);
      } else {
        entry.push(pointer);
      }
    }

    yield* seenRules.entries();
  }
}

const meta = (
  stackId: number,
  pathId: number,
  stepId: number
): GraphNodeMeta => ({ stackId, pathId, stepId });

File: packages/gbnf-experiment/data/20260909T015142Z_f8ca89ff/ported_implementation/grammar_graph/graph.py
from __future__ import annotations

from typing import Dict, Iterator, List, Optional, Sequence, Tuple

from ..utils.errors.input_parse_error import InputParseError
from ..utils.is_point_in_range import is_point_in_range
from .colorize import Colorize, colorize, no_color
from .generic_set import GenericSet
from .get_input_as_code_points import get_input_as_code_points
from .get_serialized_rule_key import get_serialized_rule_key
from .graph_node import GraphNode, GraphNodeMeta
from .graph_pointer import GraphPointer
from .rule_ref import RuleRef
from .type_guards import (
    is_range,
    is_rule_char,
    is_rule_char_excluded,
    is_rule_end,
    is_rule_ref,
)
from .types import Pointers, UnresolvedRule, ValidInput

__all__ = ["Graph"]

RootNode = Dict[int, GraphNode]


def make_pointers() -> "GenericSet[GraphPointer, str]":
    return GenericSet(lambda pointer: pointer.id)


class Graph:
    def __init__(
        self,
        grammar: str,
        stacked_rules: Sequence[Optional[Sequence[Sequence[UnresolvedRule]]]],
        root_id: int,
    ):
        self.grammar = grammar
        self._roots: Dict[int, RootNode] = {}
        self._previous_code_points: List[int] = []

        rule_refs: List[RuleRef] = []
        unique_rules: "GenericSet[UnresolvedRule, str]" = GenericSet(get_serialized_rule_key)

        for stack_id in range(len(stacked_rules)):
            stack = stacked_rules[stack_id] or []
            nodes: RootNode = {}
            for path_id in range(len(stack)):
                path = stack[path_id]
                node: Optional[GraphNode] = None
                for step_id in range(len(path) - 1, -1, -1):
                    next_node = node
                    rule = stack[path_id][step_id]
                    unique_rules.add(rule)
                    if is_rule_ref(rule):
                        rule_refs.append(rule)
                    # rules coming in may be identical but have different references.
                    # here, we ensure we always use the same reference for an identical rule.
                    # this makes future comparisons easier.
                    unique_rule = unique_rules.get(rule)
                    if unique_rule is None:
                        raise Exception("Could not get unique rule")
                    node = GraphNode(unique_rule, GraphNodeMeta(stack_id, path_id, step_id), next_node)
                if node is None:
                    raise Exception("Could not get node")
                nodes[path_id] = node

            self._roots[stack_id] = nodes

        root_node = self._roots.get(root_id)
        if root_node is None:
            raise Exception(f"Root node not found for value: {root_id}")
        self._root_node: RootNode = root_node

        for rule_ref in rule_refs:
            rule_ref.nodes = list(self._get_root_node(rule_ref.value).values())

    def _get_root_node(self, value: int) -> RootNode:
        root_node = self._roots.get(value)
        if root_node is None:
            raise Exception(f"Root node not found for value: {value}")
        return root_node

    def _get_initial_pointers(self) -> Pointers:
        pointers = make_pointers()

        root_node = self._root_node
        if root_node is None:
            raise Exception("Root node is not defined")

        for node, parent in self._fetch_nodes_for_root_node(root_node):
            pointer = GraphPointer(node, parent)
            for resolved_pointer in self._resolve_pointer(pointer):
                pointers.add(resolved_pointer)
        return pointers

    def _set_valid(self, pointers: Sequence[GraphPointer], valid: bool) -> None:
        for pointer in pointers:
            pointer.valid = valid

    def _parse(self, current_pointers: Pointers, code_point: int) -> Pointers:
        for rule, graph_pointers in self._iterate_over_pointers(current_pointers):
            if is_rule_char(rule):
                valid = False
                for possible_code_point in rule.value:
                    if valid:
                        break
                    if is_range(possible_code_point):
                        valid = is_point_in_range(code_point, possible_code_point)
                    else:
                        valid = code_point == possible_code_point
                self._set_valid(graph_pointers, valid)
            elif is_rule_char_excluded(rule):
                valid = True
                for possible_code_point in rule.value:
                    if not valid:
                        break
                    if is_range(possible_code_point):
                        valid = not is_point_in_range(code_point, possible_code_point)
                    else:
                        valid = code_point != possible_code_point
                self._set_valid(graph_pointers, valid)
            elif not is_rule_end(rule):
                raise Exception(f"Unsupported rule: {rule!r}")

        # a pointer's id is the sum of its node's id and its parent's id chain.
        # if two pointers share the same id, it means they point to the same node and have identical parent chains.
        # for the purposes of walking the graph, we only need to keep one of them.
        next_pointers = make_pointers()
        for current_pointer in current_pointers:
            for unresolved_next_pointer in current_pointer.fetch_next():
                for resolved_next_pointer in self._resolve_pointer(unresolved_next_pointer):
                    next_pointers.add(resolved_next_pointer)
        return next_pointers

    def _resolve_pointer(self, unresolved_pointer: GraphPointer) -> Iterator[GraphPointer]:
        for resolved_pointer in unresolved_pointer.resolve():
            if is_rule_ref(resolved_pointer.node.rule):
                raise Exception("Encountered a reference rule when building pointers to the graph")
            if is_rule_end(resolved_pointer.node.rule) and resolved_pointer.parent is not None:
                raise Exception("Encountered an ending rule with a parent when building pointers to the graph")
            yield resolved_pointer

    def add(self, src: ValidInput, _pointers: Optional[Pointers] = None) -> Pointers:
        pointers = _pointers if _pointers is not None else self._get_initial_pointers()
        code_points = get_input_as_code_points(src)
        for code_point_pos in range(len(code_points)):
            code_point = code_points[code_point_pos]
            pointers = self._parse(pointers, code_point)
            if pointers.size == 0:
                raise InputParseError(code_points, code_point_pos, self._previous_code_points)
        self._previous_code_points.extend(code_points)
        return pointers

    def _fetch_nodes_for_root_node(
        self,
        root_nodes: RootNode,
        parent: Optional[GraphPointer] = None,
    ) -> Iterator[Tuple[GraphNode, Optional[GraphPointer]]]:
        """Yield either the node, or if a reference rule, the referenced node.

        This is distinct from the logic in GraphPointer because that needs a rule
        ref with already-defined nodes; this function is used to _set_ those nodes.
        """
        for node in list(root_nodes.values()):
            if is_rule_ref(node.rule):
                yield from self._fetch_nodes_for_root_node(
                    self._get_root_node(node.rule.value),
                    GraphPointer(node, parent),
                )
            else:
                yield node, parent

    def print(self, pointers: Optional[Pointers] = None, colors: bool = False) -> str:
        col: Colorize = colorize if colors else no_color
        nodes: List[List[GraphNode]] = [list(node_map.values()) for node_map in self._roots.values()]
        graph_view: List[str] = []
        for root_node in nodes:
            for node in root_node:
                graph_view.append(node.print(pointers=pointers, show_position=True, colorize=col))
        return "\n" + "\n".join(graph_view)

    def __repr__(self) -> str:
        return self.print(colors=True)

    def _iterate_over_pointers(
        self, pointers: Pointers
    ) -> Iterator[Tuple[UnresolvedRule, List[GraphPointer]]]:
        # Keyed on identity: the graph canonicalizes rules on construction, so
        # pointers sharing a rule share the very same object.
        seen_rules: Dict[int, Tuple[UnresolvedRule, List[GraphPointer]]] = {}
        for pointer in pointers:
            rule = pointer.rule
            if is_rule_ref(rule):
                raise Exception("Encountered a reference rule in the graph, this should not happen")
            seen_rule = seen_rules.get(id(rule))
            if seen_rule is None:
                seen_rules[id(rule)] = (rule, [pointer])
            else:
                seen_rule[1].append(pointer)

        for rule, graph_pointers in seen_rules.values():
            yield rule, graph_pointers
